docs/architecture/adr-064-abci-2.0.md

```diff
@@ -0,0 +1,443 @@
+	for _, vote := range extCommit.Votes {
+		if !vote.SignedLastBlock || len(vote.VoteExtension) == 0 {
+			continue
+		}
+
+		valConsAddr := cmtcrypto.Address(vote.Validator.Address)
+
+		validator, err := app.validatorStore.GetValidatorByConsAddr(ctx, valConsAddr)
+		if err != nil {
+			return fmt.Errorf("failed to get validator %s for vote extension", valConsAddr)
+		}
+
+		cmtPubKey, err := validator.CmtConsPublicKey()
+		if err != nil {
+			return fmt.Errorf("failed to convert public key: %w", err)
+		}
+
+		if len(vote.ExtensionSignature) == 0 {
+			return fmt.Errorf("received a non-empty vote extension with empty signature for validator %s", valConsAddr)
+		}
+
+		cve := cmtproto.CanonicalVoteExtension{
+			Extension: vote.VoteExtension,
+			Height:    currentHeight - 1, // the vote extension was signed in the previous height
+			Round:     int64(extCommit.Round),
+			ChainId:   app.GetChainID(),
+		}
+
+		extSignBytes, err := cosmosio.MarshalDelimited(&cve)
+		if err != nil {
+			return fmt.Errorf("failed to encode CanonicalVoteExtension: %w", err)
+		}
+
+		if !cmtPubKey.VerifySignature(extSignBytes, vote.ExtensionSignature) {
+			return errors.New("received vote with invalid signature")
+		}
+
+		return nil
+	}
+}
+```
+
+Once at least 2/3 signatures, by voting power, are received and verified, the
+validator can use the vote extensions to derive additional data or come to some
+decision based on the vote extensions.
+
+> NOTE: It is very important to state, that neither the vote propagation technique
+> nor the vote extension verification mechanism described above is required for
+> applications to implement. In other words, a proposer is not required to verify
+> and propagate vote extensions along with their signatures nor are proposers
+> required to verify those signatures. An application can implement it's own
+> PKI mechanism and use that to sign and verify vote extensions.
+
+#### Vote Extension Persistence
+
+In order to make any data derived from vote extensions persistent, we propose to
+to allow application developers to "merge" the `processProposalState` into the
+`finalizeState`, such that when processing transactions during `FinalizeBlock`,
+the base state includes any state changes from processing votes extensions
+(see [`FinalizeBlock`](#finalizeblock-1) below).
+
+However, we DO NOT want to pollute `finalizeState` with any state changes from
+verifying transactions during `ProcessProposal`, so we will introduce a new API,
+`ResetProcessProposalState`, that will allow applications to essentially reset
+the `ProcessProposal` state, effectively allowing any state transitions after to
+be committed.
+
+What this means is that any explicit state changes in a `ProcessProposal` handler
+using the injected `processProposalState.Context` will be written to state! An
+application must be careful to execute `ResetProcessProposalState` where and when
+appropriate.
```

Comment:
To confirm my understanding, is the following correct?

1. ProcessProposalHandler: does a bunch of state modifications
2. ProcessProposalHandler: if `ResetProcessProposalState` is called, then all the modifications made in 1 is reverted. If not, then all the modifications will be passed onto the next step, which is `FinalizeBlock` which will update the actual state
3. Since `SetVoteExtResult` is called after the `Reset`, the VoteExtResult will be available in `FinalizeBlock` state?

Reply:
Bingo!

Reply:
This was the cleanest way I could think of doing it.

Reply:
got it, thanks for clarifying. feel free to resolve this comment!